### PR TITLE
[Qwen2Audio] handle input ids expansion during processing

### PR DESCRIPTION
**Pull request by eustlb**
# What does this PR do?

This PR adds input_id expansion to the processor. 

Before: 
1. embedding of the input_ids`"<|audio_bos|><|AUDIO|><|audio_eos|>Generate the caption in English:"` →  `inputs_embeds` tensor
2. creation of a new embedding tensor [here](https://github.com/huggingface/transformers/blob/a821b9c7ab6b0ce4a4557be770cba5946df5c322/src/transformers/models/qwen2_audio/modeling_qwen2_audio.py#L1051C8-L1053C10), taking into account the number of audio tokens to be merged with inputs_embeds
3. replacement of audio embedding vectors 

Now:
1. expansion of input ids at the processing stage: `"<|audio_bos|><|AUDIO|><|audio_eos|>Generate the caption in English:"` →  `"<|audio_bos|><|AUDIO|>...<|AUDIO|><|audio_eos|>Generate the caption in English:"` with as many` <|AUDIO|>` tokens as the number of audio embedded vectors after encoding and projection
2. creation of `inputs_embeds`, directly with the correct shape
3. replacement of the audio embedding vectors

This approach allows to remove the unnecessary step of creating a new embedding tensor in the model forward. Moreover, it simplifies the code as correct padding is directly handled at the processing stage.

### CODE REVIEW

File: src/transformers/models/qwen2_audio/processing_qwen2_audio.py [hunk 1]
```diff
@@ -97,6 +101,26 @@ def __call__(
             audio_inputs["feature_attention_mask"] = audio_inputs.pop(
                 "attention_mask"
             )  # rename attention_mask to prevent conflicts later on
+
+            expanded_text = []
```
Comment (ArthurZucker) [inline src/transformers/models/qwen2_audio/processing_qwen2_audio.py:105]: IMO allowing people to just pass `Hey<audio>` instead of `Hey<audio_bos><audio><audio_eos>` makes it even simpler! (cf the PR description so I am not 100 sure!)

File: src/transformers/models/qwen2_audio/modeling_qwen2_audio.py [hunk 2]
```diff
@@ -1197,9 +1197,39 @@ def forward(
...
                 audio_features = self.multi_modal_projector(selected_audio_feature)
 
-                inputs_embeds, attention_mask, labels, position_ids, _ = self._merge_input_ids_with_audio_features(
-                    audio_features, audio_output_lengths, inputs_embeds, input_ids, attention_mask, labels
-                )
+                # if we have consecutive audio tokens, then it means we expanded input_ids in processing
+                audio_tokens = input_ids == self.config.audio_token_index
+                legacy_processing = (audio_tokens[:, :-1] & audio_tokens[:, 1:]).sum() == 0
+
+                if legacy_processing:
```
Comment (ArthurZucker) [inline src/transformers/models/qwen2_audio/modeling_qwen2_audio.py:1204]: could be an argument to merge inputs ids with audio features to do this in `_merge_input_ids_with_audio_features` as people have requested customization of this funciton

File: src/transformers/models/qwen2_audio/processing_qwen2_audio.py [hunk 3]
```diff
@@ -97,6 +101,26 @@ def __call__(
...
+                    audio_length = audio_lengths.pop(0)
+                    input_length = (audio_length - 1) // 2 + 1
+                    num_audio_tokens = (input_length - 2) // 2 + 1
+                    replace_str.append(self.audio_token * num_audio_tokens)
+                    sample = sample.replace(self.audio_token, "<placeholder>", 1)
+
+                while "<placeholder>" in sample:
+                    sample = sample.replace("<placeholder>", replace_str.pop(0), 1)
+                expanded_text.append(sample)
+            text = expanded_text
```
Comment (ArthurZucker) [inline src/transformers/models/qwen2_audio/processing_qwen2_audio.py:119]: processor needs to error out gracefully when you don't have a correct number of text token vs audio features: https://github.com/huggingface/transformers/blob/af28ac4a0a7807a190e20ee098cd2377324da9fe/src/transformers/models/pixtral/processing_pixtral.py#L216-L218 

Not sure while make it easier to understand what's happening vs add the lengths to a list and zipping. I prefer list + zip for explicity but not a must! 🤗

**Review by ArthurZucker — COMMENTED** [review]
Nice!

File: src/transformers/models/qwen2_audio/modeling_qwen2_audio.py [hunk 4]
```diff
@@ -1197,9 +1197,39 @@ def forward(
...
+                        "Expanding inputs for audio tokens in Qwen2Audio should be done in processing."
+                    )
+                    inputs_embeds, attention_mask, labels, position_ids, _ = self._merge_input_ids_with_audio_features(
+                        audio_features, audio_output_lengths, inputs_embeds, input_ids, attention_mask, labels
+                    )
+                else:
+                    num_audios, max_audio_tokens, embed_dim = audio_features.shape
+                    audio_features_mask = torch.arange(max_audio_tokens).expand(num_audios, max_audio_tokens).to(
+                        audio_output_lengths.device
+                    ) < audio_output_lengths.unsqueeze(1)
```
[the remark this answers is not in the document]
Comment (eustlb) [reply src/transformers/models/qwen2_audio/modeling_qwen2_audio.py]: addressed in [bd4ec1c](https://github.com/huggingface/transformers/pull/35534/commits/bd4ec1c743da6185abc9d0238b16138a51aac5a8)

File: src/transformers/models/qwen2_audio/modeling_qwen2_audio.py [hunk 5]
```diff
@@ -1197,9 +1197,39 @@ def forward(
...
+                    if n_audio_tokens != n_audio_features:
+                        raise ValueError(
+                            f"Audio features and audio tokens do not match: tokens: {n_audio_tokens}, features {n_audio_features}"
+                        )
+                    special_audio_mask = (
+                        (input_ids == self.config.audio_token_index)
+                        .unsqueeze(-1)
+                        .expand_as(inputs_embeds)
+                        .to(inputs_embeds.device)
+                    )
```
[the remark this answers is not in the document]
Comment (eustlb) [reply src/transformers/models/qwen2_audio/modeling_qwen2_audio.py]: addressed in [bd4ec1c](https://github.com/huggingface/transformers/pull/35534/commits/bd4ec1c743da6185abc9d0238b16138a51aac5a8)

File: src/transformers/models/qwen2_audio/modeling_qwen2_audio.py (hunk 2 above)
Comment (eustlb) [reply src/transformers/models/qwen2_audio/modeling_qwen2_audio.py:1204]: It is not done this way for vision model (e.g. llava) where it has been kept separated from `_merge_input_ids_with_audio_features`, likely to directly suppress `_merge_input_ids_with_audio_features` when `legacy_processing` is deprecated. I would do the same here for audio, WDYT?

File: src/transformers/models/qwen2_audio/processing_qwen2_audio.py (hunk 1 above)
Comment (eustlb) [reply src/transformers/models/qwen2_audio/processing_qwen2_audio.py:129]: I've added in fa85ac401bc91c6faf326a27b11a55bc341fa47d the possibility to pass only `Hey<audio>` for simplicy + added an example in 4569047a908f48fdd21d9e91c11ff5d76aa9e882

File: tests/models/qwen2_audio/test_modeling_qwen2_audio.py [hunk 6]
```diff
@@ -280,7 +284,11 @@ def test_small_model_integration_test_single(self):
         )
         self.assertTrue(torch.equal(inputs["input_ids"], EXPECTED_INPUT_IDS))
 
-        EXPECTED_DECODED_TEXT = "<|im_start|>system\nYou are a helpful assistant.<|im_end|>\n<|im_start|>user\nAudio 1: <|audio_bos|><|AUDIO|><|audio_eos|>\nWhat's that sound?<|im_end|>\n<|im_start|>assistant\nIt is the sound of glass breaking.<|im_end|>"
+        EXPECTED_DECODED_TEXT = (
+            "<|im_start|>system\nYou are a helpful assistant.<|im_end|>\n<|im_start|>user\nAudio 1: <|audio_bos|>"
+            + "<|AUDIO|>" * 101
+            + "<|audio_eos|>\nWhat's that sound?<|im_end|>\n<|im_start|>assistant\nIt is the sound of glass breaking.<|im_end|>"
+        )
```
[the remark this answers is not in the document]
Comment (eustlb) [reply tests/models/qwen2_audio/test_modeling_qwen2_audio.py:292]: test added in 09870d849db69e597ebef759f8a1567fa093fc19

File: src/transformers/models/qwen2_audio/processing_qwen2_audio.py (hunk 3 above)
Comment (eustlb) [reply src/transformers/models/qwen2_audio/processing_qwen2_audio.py]: I've added the error in 71ce83f54d5afc8ed77d0b771b849125c6ce0f0a. Nevertheless, if you agree, I would keep the while loop: it makes things simpler as the zip would not be direct since we can have `[text1, text2]` with `[audio1, audio2, audio3, audio4]` where `audio1` and `audio2` are for `text1` and `audio3` and `audio4` for `text2`

File: src/transformers/models/qwen2_audio/modeling_qwen2_audio.py (hunk 2 above)
Comment (ArthurZucker) [reply src/transformers/models/qwen2_audio/modeling_qwen2_audio.py:1204]: Yep up to you. I got confused with https://github.com/huggingface/transformers/issues/33696 ! So all good

File: src/transformers/models/qwen2_audio/processing_qwen2_audio.py (hunk 3 above)
Comment (ArthurZucker) [reply src/transformers/models/qwen2_audio/processing_qwen2_audio.py]: yep no worries!

File: tests/models/qwen2_audio/test_modeling_qwen2_audio.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply tests/models/qwen2_audio/test_modeling_qwen2_audio.py:292]: thanks

File: tests/models/qwen2_audio/test_modeling_qwen2_audio.py [hunk 7]
```diff
@@ -280,13 +284,64 @@ def test_small_model_integration_test_single(self):
...
+                    5112,
+                    30,
+                    151645,
+                    198,
+                    151644,
+                    77091,
+                    198,
+                ]
+            ]
+        )
```
Comment (ArthurZucker) [inline tests/models/qwen2_audio/test_modeling_qwen2_audio.py:339]: use # fmt: skip

**Review by ArthurZucker — APPROVED** [review]
Thanks 🤗
